### PR TITLE
Implement Serpentine Helmet (fixed)

### PR DESCRIPTION
**Pull request by InfernalPlank**
This is a remake of https://github.com/geheur/weapon-charges/pull/8 but with the code refactored and properly ~~working~~ tracking.
Additionally, the number and size of commits has been reduced.

This PR implements the [Serpentine Helmet](https://oldschool.runescape.wiki/w/Serpentine_helm) into the list of items tracked by the plugin.

### CODE REVIEW

File: src/main/java/com/weaponcharges/WeaponChargesItemOverlay.java [hunk 1]
```diff
@@ -119,6 +126,39 @@ public void renderItemOverlay(Graphics2D graphics, int itemId, WidgetItem itemWi
...
+				if (displayWhen == DisplayWhen.NEVER && !plugin.isShowChargesKeyIsDown()) return;
+				NumberFormat df = new DecimalFormat("##0.0");
+				float scalesLeftPercent = (float) charges / chargedWeapon.rechargeAmount;
+				df.setRoundingMode(RoundingMode.DOWN);
+				if (config.serpentine_helm_DisplayStyle() == SerpModes.SCALES) {
+					topText.setText(charges.toString());
+				} else if (config.serpentine_helm_DisplayStyle() == SerpModes.PERCENT) {
+					topText.setText(df.format((scalesLeftPercent * 100)) + "%");
+					if (df.format((scalesLeftPercent * 100)).equals(df.format(0.0)) && charges > 0) {
+						topText.setText("1.0%");
```
Comment (geheur) [inline src/main/java/com/weaponcharges/WeaponChargesItemOverlay.java:147]: should be 0.1%.

Comment (geheur) [reply src/main/java/com/weaponcharges/WeaponChargesItemOverlay.java:147]: or whatever the smallest non-0-looking value is.

File: src/main/java/com/weaponcharges/WeaponChargesPlugin.java [hunk 2]
```diff
@@ -214,15 +214,35 @@ private void enableDevMode()
...
-		if (crystalHelm != null) addCharges(crystalHelm, -1, false);
-		ChargedWeapon crystalBody = getEquippedChargedWeapon(EquipmentInventorySlot.BODY);
-		if (crystalBody != null) addCharges(crystalBody, -1, false);
-		ChargedWeapon crystalLegs = getEquippedChargedWeapon(EquipmentInventorySlot.LEGS);
-		if (crystalLegs != null) addCharges(crystalLegs, -1, false);
+		Hitsplat.HitsplatType hitType = e.getHitsplat().getHitsplatType();
+		ChargedWeapon helm = getEquippedChargedWeapon(EquipmentInventorySlot.HEAD);
+		if (helm == ChargedWeapon.SERPENTINE_HELM) {
+			if (hitType == Hitsplat.HitsplatType.DAMAGE_ME || hitType == Hitsplat.HitsplatType.BLOCK_ME) {
+				if (client.getTickCount() - lastDegradedHitsplatTick >= 90) {
```
Comment (geheur) [inline src/main/java/com/weaponcharges/WeaponChargesPlugin.java:225]: I think > is correct, not >=, unless you have testing that says otherwise?

Comment (geheur) [reply src/main/java/com/weaponcharges/WeaponChargesPlugin.java:225]: I've never seen a 90-tick gap, smallest I've seen is 91.

File: src/main/java/com/weaponcharges/WeaponChargesPlugin.java [hunk 3]
```diff
@@ -214,15 +214,35 @@ private void enableDevMode()
...
-		ChargedWeapon crystalHelm = getEquippedChargedWeapon(EquipmentInventorySlot.HEAD);
-		if (crystalHelm != null) addCharges(crystalHelm, -1, false);
-		ChargedWeapon crystalBody = getEquippedChargedWeapon(EquipmentInventorySlot.BODY);
-		if (crystalBody != null) addCharges(crystalBody, -1, false);
-		ChargedWeapon crystalLegs = getEquippedChargedWeapon(EquipmentInventorySlot.LEGS);
-		if (crystalLegs != null) addCharges(crystalLegs, -1, false);
+		Hitsplat.HitsplatType hitType = e.getHitsplat().getHitsplatType();
+		ChargedWeapon helm = getEquippedChargedWeapon(EquipmentInventorySlot.HEAD);
+		if (helm == ChargedWeapon.SERPENTINE_HELM) {
+			if (hitType == Hitsplat.HitsplatType.DAMAGE_ME || hitType == Hitsplat.HitsplatType.BLOCK_ME) {
```
Comment (geheur) [inline src/main/java/com/weaponcharges/WeaponChargesPlugin.java:224]: Could use a comment explaining what DAMAGE_ME and BLOCK_ME and why they are appropriate and sufficient, as the javadoc is not very clear.

Comment (geheur) [reply src/main/java/com/weaponcharges/WeaponChargesPlugin.java:224]: idc too much though.

File: src/main/java/com/weaponcharges/ChargedWeapon.java [hunk 4]
```diff
@@ -623,6 +623,54 @@ periodic updates (periodic, empty):
 			ChargesMessage.matcherGroupChargeMessage("Your crystal legs has ([\\d,]+) charges remaining.", 1)
 		)
 	),
+	/* Serpentine Helmet:
```
Comment (geheur) [inline src/main/java/com/weaponcharges/ChargedWeapon.java:626]: An explanation of the serp helm degradation mechanics, as well as the likely deficiencies of the current approach visavis ranged attacks would be nice here.

Comment (geheur) [reply src/main/java/com/weaponcharges/ChargedWeapon.java:626]: Also the barrows armor timer sharing.

File: src/test/java/com/weaponcharges/ConfigGenerator.java [hunk 5]
```diff
@@ -4,7 +4,7 @@
 
 public class ConfigGenerator
 {
-	static int i = 0;
+	static int i = -1;
```
Comment (geheur) [inline src/test/java/com/weaponcharges/ConfigGenerator.java:7]: why did you change this.

File: src/main/java/com/weaponcharges/WeaponChargesItemOverlay.java [hunk 6]
```diff
@@ -66,7 +69,11 @@ public void renderItemOverlay(Graphics2D graphics, int itemId, WidgetItem itemWi
 					boolean isLowCharge = charges <= chargedWeapon.getLowCharge(config);
 					if (!isLowCharge && displayWhen == DisplayWhen.LOW_CHARGE && !plugin.isShowChargesKeyIsDown()) break;
 
-					topText.setText(String.valueOf(charges));
+					if (chargedWeapon == ChargedWeapon.SERPENTINE_HELM && charges == 0) {
```
Comment (geheur) [inline src/main/java/com/weaponcharges/WeaponChargesItemOverlay.java:72]: Make this empty thing a config option that affects all weapons, replacing "0" on all uncharged versions with "Empty".

File: src/main/java/com/weaponcharges/WeaponChargesItemOverlay.java (hunk 1 above)
Comment (geheur) [reply src/main/java/com/weaponcharges/WeaponChargesItemOverlay.java:147]: actually, is a non-0 value even capable of producing 0.0%? have you tested that?

File: src/main/java/com/weaponcharges/WeaponChargesItemOverlay.java [hunk 7]
```diff
@@ -119,6 +126,39 @@ public void renderItemOverlay(Graphics2D graphics, int itemId, WidgetItem itemWi
...
+				topText.setText("??.?%");
+			} else {
+				if (displayWhen == DisplayWhen.NEVER && !plugin.isShowChargesKeyIsDown()) return;
+				NumberFormat df = new DecimalFormat("##0.0");
+				float scalesLeftPercent = (float) charges / chargedWeapon.rechargeAmount;
+				df.setRoundingMode(RoundingMode.DOWN);
+				if (config.serpentine_helm_DisplayStyle() == SerpModes.SCALES) {
+					topText.setText(charges.toString());
+				} else if (config.serpentine_helm_DisplayStyle() == SerpModes.PERCENT) {
+					topText.setText(df.format((scalesLeftPercent * 100)) + "%");
```
Comment (geheur) [inline src/main/java/com/weaponcharges/WeaponChargesItemOverlay.java:145]: format code is still duped.

File: src/main/java/com/weaponcharges/WeaponChargesItemOverlay.java (hunk 1 above)
Comment (geheur) [reply src/main/java/com/weaponcharges/WeaponChargesItemOverlay.java:147]: oh I see, that's how I do my blowpipe code. nevermind.

So make sure that when you remove the duplicated code you remove it from the blowpipe code also.

**Review by geheur — CHANGES_REQUESTED** [review]
Haven't tested in-game yet, this review is just looking at the code.

File: src/main/java/com/weaponcharges/WeaponChargesItemOverlay.java (hunk 1 above)
Comment (RHQ-Rusty) [reply src/main/java/com/weaponcharges/WeaponChargesItemOverlay.java:147]: > So make sure that when you remove the **duplicated code** you remove it from the blowpipe code also.

By "duplicated code" are you referring to the entire conditional check there?
```java
if (df.format((scalesLeftPercent * 100)).equals(df.format(0.0)) && charges > 0) {
	topText.setText("1.0%");
}
```
I can remove this from Serpentine and Blowpipe.

File: src/main/java/com/weaponcharges/ChargedWeapon.java (hunk 4 above)
Comment (RHQ-Rusty) [reply src/main/java/com/weaponcharges/ChargedWeapon.java:626]: Implemented in commit 169e8e8360554165fa4eead7377f82ae5817bfc3.

File: src/test/java/com/weaponcharges/ConfigGenerator.java (hunk 5 above)
Comment (RHQ-Rusty) [reply src/test/java/com/weaponcharges/ConfigGenerator.java:7]: This is a change made for `ConfigGenerator.java`'s new functionality (support for skipping serp display style config item) and does not break anything. It works as intended.

File: src/main/java/com/weaponcharges/WeaponChargesItemOverlay.java (hunk 1 above)
Comment (RHQ-Rusty) [reply src/main/java/com/weaponcharges/WeaponChargesItemOverlay.java:147]: De-duplicated; leaving in until it's confirmed to not be needed. It is a nice edge-case to have covered.

File: src/main/java/com/weaponcharges/WeaponChargesItemOverlay.java (hunk 7 above)
Comment (RHQ-Rusty) [reply src/main/java/com/weaponcharges/WeaponChargesItemOverlay.java:145]: Fixed.

File: src/main/java/com/weaponcharges/WeaponChargesItemOverlay.java (hunk 6 above)
Comment (RHQ-Rusty) [reply src/main/java/com/weaponcharges/WeaponChargesItemOverlay.java:72]: Finished. Pushing in a moment.

File: src/main/java/com/weaponcharges/WeaponChargesPlugin.java (hunk 2 above)
Comment (RHQ-Rusty) [reply src/main/java/com/weaponcharges/WeaponChargesPlugin.java:225]: I will remove the = and we can revisit this if desync is apparent as a result.

File: src/main/java/com/weaponcharges/WeaponChargesPlugin.java (hunk 3 above)
Comment (RHQ-Rusty) [reply src/main/java/com/weaponcharges/WeaponChargesPlugin.java:224]: Left a note explaining the splats.

**Comment by RHQ-Rusty** [comment]
All requested changes have been made! Let me know if anything else needs done.
